### DISCUSSION TITLE
File names in `icon_cache` leaking vault content

### DISCUSSION THREAD

**Discussion by bentolor** [dani-garcia/vaultwarden#3987] [Q&A] [answered]
I'm not sure if this is due to the Bitwarden client itself or specific to the Vaultwarden implementation. But looking at the `icon_cache` content exposes the websites I have passwords for. 

I perceive that information compromising the confidentiality of my vault. For example it exposes non-public host names.

My questions

a) Is this due to the Bitwarden client protocol itself? Can somebody explain where it applies?
b) I'm rather suprised, not to say shocked, about this leackage of vault content. I thought Bitwarden would not share the same design flaw as the plain URLs in Lastpass. 

Any comments?
Benjamin

**Comment by BlackDex** [comment]
It isn't useful to encrypt or obfuscate those file, we still need to serve them unencrypted and need to know which file to use.
Since Bitwarden can only send the unencrypted plain/text host/domain we also need to match that.

Hashing those files would not matter, since people can create there own hash tables and match them.
If you do not want this, disable icons, or use the redirect options.

**Reply by bentolor** [reply]
Hi @BlackDex ! Thanks for your prompt and helpful reply. It helped me to verify, that it's due to Bitwardens client which asks/requests the content via plaintext URLs. I also found now the relevant page: https://bitwarden.com/help/website-icons/

I'm on the same page with you, that hashing would only be cosmetics and security theater.

Bitwarden even confirms this confidentiality loss on their own page: 

> The icon server endpoints do not log or collect any information regarding icon image requests, however this is something **you would have to take our word** for since we have no way to demonstrate this publicly.

I'm still in awe by this design. For illustration: Just imagine stumbling over following content in `icon_cache` in your vaultwarden instance only used by yourself and your partner:

```
grindr.com
tinder.com
adam4adam.com
smartrecovery.org
```

*Fun fact:* Your are not aware of any drug addiction of your partner, nor did you signup for all those sites yourself.

Discuss!

**Reply by bentolor** [reply]
As a proposal: How about to puting a note / hint on the according vaultwarden option? I just found out, that vaultwarden already provides an according option (great!).

If it was me, I'd disable it by default. Though this probably would not stop the bitwarden client itself to ask/request the URLs, hence still exposing vault content i.e. as part of access logs etc.

**Reply by BlackDex** [reply, marked as answer]
Most people actually want those favicons, so, disabling by default is probably not a good option.

You can also disable it on the client side, as the documentation describes. That will prevent the clients from calling the URL at all, but you have to do that for all clients separately.

An other option would be to change the first URL in the password item to something totally unrelated, like `disney.com` or so, and set that URL's ` Match Detection` to never, and set the second URL to the correct item.

This will cause the icon to be from Disney, but it will never match Disney.com.

I use this for my GPG Keys for example, it has a nice GPG icon, but will never match that site.

**Reply by bentolor** [reply]
Very valueable hints! Especially how to "tweak" custom favicons. Thanks a lot for the inspiration here, @BlackDex !

I'd see only client-side disabling as a solid workaround. And still: Only a workaround.

Clearly, this is an upstream issue. I've mixed feelings, if I do want to spare the energy to discuss and raise this upstream. The rebuttal of "works as documented" is obviously to be expected.

So probably my most energy-saving "solution" would be to
a) tell my partner about this behaviour and
b) refrain from snooping to much in the `icon_cache` dir

But honestly: That still is no in sync with the expectations I have for the confidentiality of a password vault.

**Reply by dani-garcia** [reply]
Ultimately this problem is only an issue on installations with a small number of users, once you have more than a couple of users it gets almost impossible to correlate which user visited which site based on the icon cache alone, as the cache is globally shared. So I can imagine why this is not a big problem for Bitwarden. There is still the trust issue that you don't know if  the server logs anything when doing a request like `GET /icons/sensitive_domain.com/icon.png`, but there's not much to do there, as the server needs access to the plaintext domain to fetch the icon anyway.

If Bitwarden synced the settings between all the clients, this would be less of an issue, as you'd only need to configure the icon disabling once per account, and it's something that we could force for all users from Vaultwarden's side, but at least for the short and medium term that feature is not on the roadmap.

I can see how this can cause an issue for single user or very low user installations, it's just that realistically there is nothing much that we can do other than apply some obscurity layers on top. That said, maybe it's worth it to just do some basic hashing of the domain to avoid accidental snooping.
